variables.tf

```diff
@@ -1,41 +1,133 @@
+    If not provided, will be derived from the `null-label.context` passed in.
+    If `create_before_destroy` is true, will be used as a name prefix.
   EOT
 }
 
+variable "security_group_description" {
+  type        = string
+  default     = "Managed by Terraform"
+  description = <<-EOT
+    The description to assign to the created Security Group.
+    Warning: Changing the description causes the security group to be replaced, which requires everything
+    associated with the security group to be replaced, which can be very disruptive.
+    EOT
+}
+
+variable "create_before_destroy" {
+  type        = bool
+  default     = false
+  description = <<-EOT
+    Set `true` to enable terraform `create_before_destroy` behavior on the created security group.
+    We recommend setting this `true` on new security groups, but default it to `false` because `true`
+    will cause existing security groups to be replaced.
+    Note that changing this value will also cause the security group to be replaced.
+    EOT
+}
+
+variable "allow_all_egress" {
+  type        = bool
+  default     = false
+  description = <<-EOT
+    A convenience that adds to the rules specified elsewhere a rule that allows all egress.
+    If this is false and no egress rules are specified via `rules` or `rule-matrix`, then no egress will be allowed.
+    EOT
+}
+
 variable "rules" {
   type        = list(any)
-  default     = null
+  default     = []
   description = <<-EOT
-    A list of maps of Security Group rules. 
-    The values of map is fully complated with `aws_security_group_rule` resource. 
+    A list of maps of Security Group rules.
+    The keys and values of the maps are fully compatible with the `aws_security_group_rule` resource, except
+    for `security_group_id` which will be ignored.
     To get more info see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule .
   EOT
 }
+
+variable "rule_matrix" {
+  # rule_matrix is independent of the `rules` input.
+  # Only the rules specified in the `rule_matrix` object are applied to the subjects specified in `rule_matrix`.
+  #  Schema:
+  #  {
+  #    # these top level lists define all the subjects to which rule_matrix rules will be applied
+  #    source_security_group_ids = list of source security group IDs to apply all rules to
+  #    cidr_blocks = list of ipv4 CIDR blocks to apply all rules to
+  #    ipv6_cidr_blocks = list of ipv6 CIDR blocks to apply all rules to
+  #    prefix_list_ids = list of prefix list IDs to apply all rules to
+  #    self = # set "true" to apply the rules to the created or existing security group
+  #
+  #    # each rule in the rules list will be applied to every subject defined above
+  #    rules = [{
+  #      type = "ingress"
+  #      from_port = 433
+  #      to_port = 433
+  #      protocol = "tcp"
+  #      description = "Allow HTTPS ingress"
+  #    }]
+
+  type        = any
```

Comment:
While not a bad idea, we do not want to enforce type constraints here because this is a core module expected to evolve over time. We learned this lesson with [null-label](https://github.com/cloudposse/terraform-null-label/). By allowing any input, we give ourselves more ability to add new features while maintaining backward compatibility.

Reply:
Thanks, makes sense